services/wallet/router.go

```diff
@@ -704,21 +722,26 @@ func (r *Router) suggestedRoutes(
 					}
 					gasLimit := uint64(0)
 					if sendType.isTransfer() {
-						gasLimit, err = bridge.EstimateGas(network, dest, addrFrom, addrTo, token, amountIn)
+						gasLimit, err = bridge.EstimateGas(network, dest, addrFrom, addrTo, token, toToken, amountIn)
 						if err != nil {
 							continue
 						}
 					} else {
 						gasLimit = sendType.EstimateGas(r.s, network, addrFrom, tokenID)
 					}
 
-					approvalRequired, approvalAmountRequired, approvalGasLimit, l1ApprovalFee, approvalContractAddress, err := r.requireApproval(ctx, sendType, bridge, addrFrom, network, token, amountIn)
+					approvalContractAddress := bridge.GetContractAddress(network, token)
+					if approvalContractAddress == nil {
+						continue
+					}
+
+					approvalRequired, approvalAmountRequired, approvalGasLimit, l1ApprovalFee, err := r.requireApproval(ctx, sendType, *approvalContractAddress, addrFrom, network, token, amountIn)
 					if err != nil {
 						continue
 					}
 
-					var l1GasFeeWei uint64 = 0
-					if sendType.needL1Fee() {
+					var l1GasFeeWei uint64
+					if sendType != Swap && sendType.needL1Fee() {
```

Comment:
in case of optimism, we still need this right?

Reply:
Also not sure about it, most likely yes, but I added that condition because of an inability to build a tx for swapping erc20 tokens without executed approval tx.